### PR TITLE
[SPARK-43461][BUILD] Skip compiling useless files when making distribution

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR add more skip properties when making distribution:
- `-Dmaven.javadoc.skip=true` to skip generating javadoc
- `-Dmaven.scaladoc.skip=true` to skip generating scaladoc. Please see: https://davidb.github.io/scala-maven-plugin/doc-jar-mojo.html#skip
- `-Dmaven.source.skip` to skip generating sources.jar
- `-Dcyclonedx.skip=true` to skip making bom. Please see: https://cyclonedx.github.io/cyclonedx-maven-plugin/makeBom-mojo.html#skip

### Why are the changes needed?

Reduce time spent on making distribution.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manual test:
```sh
./dev/make-distribution.sh --tgz --pip -Phadoop-3 -Phive -Phive-thriftserver -Pyarn -Phadoop-provided
```

Before this PR | After this PR
-- | --
43 min total from scheduled to completion | 23 min total from scheduled to completion

### CODE REVIEW

**Comment by LuciferYang** [comment]
I found some downloading log when run `dev/make-distribution.sh --tgz` with this pr

image


is this expected

**Comment by LuciferYang** [comment]
@wangyum I think we should not skip building the test jar. I test the following commands:

```
build/mvn versions:set -DnewVersion=3.5.0.1
dev/make-distribution.sh --tgz 
```

then the build will failed as follows:

```
[INFO] -----------------< org.apache.spark:spark-sketch_2.12 >-----------------
[INFO] Building Spark Project Sketch 3.5.0.1                             [3/30]
[INFO]   from common/sketch/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
Downloading from gcs-maven-central-mirror: https://maven-central.storage-download.googleapis.com/maven2/org/apache/spark/spark-tags_2.12/3.5.0.1/spark-tags_2.12-3.5.0.1-tests.jar
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/spark/spark-tags_2.12/3.5.0.1/spark-tags_2.12-3.5.0.1-tests.jar
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spark Project Parent POM 3.5.0.1:
[INFO] 
[INFO] Spark Project Parent POM ........................... SUCCESS [  3.746 s]
[INFO] Spark Project Tags ................................. SUCCESS [  4.028 s]
[INFO] Spark Project Sketch ............................... FAILURE [01:16 min]
[INFO] Spark Project Local DB ............................. SKIPPED
[INFO] Spark Project Networking ........................... SKIPPED
[INFO] Spark Project Shuffle Streaming Service ............ SKIPPED
[INFO] Spark Project Unsafe ............................... SKIPPED
[INFO] Spark Project Common Utils ......................... SKIPPED
[INFO] Spark Project Launcher ............................. SKIPPED
[INFO] Spark Project Core ................................. SKIPPED
[INFO] Spark Project ML Local Library ..................... SKIPPED
[INFO] Spark Project GraphX ............................... SKIPPED
[INFO] Spark Project Streaming ............................ SKIPPED
[INFO] Spark Project Catalyst ............................. SKIPPED
[INFO] Spark Project SQL .................................. SKIPPED
[INFO] Spark Project ML Library ........................... SKIPPED
[INFO] Spark Project Tools ................................ SKIPPED
[INFO] Spark Project Hive ................................. SKIPPED
[INFO] Spark Project REPL ................................. SKIPPED
[INFO] Spark Project Assembly ............................. SKIPPED
[INFO] Kafka 0.10+ Token Provider for Streaming ........... SKIPPED
[INFO] Spark Integration for Kafka 0.10 ................... SKIPPED
[INFO] Kafka 0.10+ Source for Structured Streaming ........ SKIPPED
[INFO] Spark Project Examples ............................. SKIPPED
[INFO] Spark Integration for Kafka 0.10 Assembly .......... SKIPPED
[INFO] Spark Avro ......................................... SKIPPED
[INFO] Spark Project Connect Common ....................... SKIPPED
[INFO] Spark Project Connect Server ....................... SKIPPED
[INFO] Spark Project Connect Client ....................... SKIPPED
[INFO] Spark Protobuf ..................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:24 min
[INFO] Finished at: 2023-05-12T00:15:34+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project spark-sketch_2.12: Could not resolve dependencies for project org.apache.spark:spark-sketch_2.12:jar:3.5.0.1: Could not transfer artifact org.apache.spark:spark-tags_2.12:jar:tests:3.5.0.1 from/to central (https://repo.maven.apache.org/maven2): transfer failed for https://repo.maven.apache.org/maven2/org/apache/spark/spark-tags_2.12/3.5.0.1/spark-tags_2.12-3.5.0.1-tests.jar: Connect to repo.maven.apache.org:443 [repo.maven.apache.org/151.101.76.215, repo.maven.apache.org/2a04:4e42:12:0:0:0:0:215] failed: No route to host (connect failed) -> [Help 1]
```

the test jar `spark-tags_2.12-3.5.0.1-tests.jar` cannot be downloaded

**Comment by wangyum** [comment]
Thank you @LuciferYang. It's not expected.

**Comment by pan3793** [comment]
`-Dskip=true` also skips building the test jar, in Kyuubi, we introduce another maven property `<maven.scaladoc.skip>false</maven.scaladoc.skip>` to control the scaladoc

**Review by pan3793 — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

File: dev/make-distribution.sh [hunk 1]
```diff
@@ -166,7 +166,12 @@ export MAVEN_OPTS="${MAVEN_OPTS:--Xss128m -Xmx4g -XX:ReservedCodeCacheSize=128m}
 # Store the command as an array because $MVN variable might have spaces in it.
 # Normal quoting tricks don't work.
 # See: http://mywiki.wooledge.org/BashFAQ/050
-BUILD_COMMAND=("$MVN" clean package -DskipTests $@)
+BUILD_COMMAND=("$MVN" clean package \
+    -DskipTests \
+    -Dmaven.javadoc.skip=true -Dmaven.scaladoc.skip=true \
```
Comment (dongjoon-hyun) [inline dev/make-distribution.sh:171]: Although these are all about the doc, could you split this into two lines?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (only one minor comments). Thank you, @wangyum and all.

File: dev/make-distribution.sh (hunk 1 above)
Comment (wangyum) [reply dev/make-distribution.sh]: ok

**Comment by LuciferYang** [comment]
Thanks @wangyum @dongjoon-hyun @pan3793 . Merged to master.

**Pull request closed without merging**